tb_plugin/torch_tb_profiler/profiler/range_lib.py

```diff
@@ -0,0 +1,144 @@
+from enum import IntEnum
+
+
+EndpointTypes = IntEnum('EndpointTypes', ['START', 'END'], start=0)
+
+
+class EndPoint(object):
+    def __init__(self, ep_time, ep_pt_type, ep_value):
+        self.time = ep_time
+        self.pt_type = ep_pt_type
+        self.value = ep_value
+
+
+# src_ranges: item of ((start_time, end_time), value)
+def merge_ranges_with_value(src_ranges):
+    merged_ranges = []
+    if len(src_ranges) > 0:
+        # Build tuple of (time, type, value)
+        endpoints = []
+        for r in src_ranges:
+            endpoints.append(EndPoint(r[0][0], EndpointTypes.START, r[1]))
+            endpoints.append(EndPoint(r[0][1], EndpointTypes.END, r[1]))
+        endpoints.sort(key=lambda x: [x.time, int(x.pt_type)])  # Make START in front of END if equal on time.
+
+        last_endpoint_time = endpoints[0].time
+        last_value = endpoints[0].value
+        for i in range(1, len(endpoints)):
+            ep = endpoints[i]
+            if ep.time > last_endpoint_time and last_value > 0.0:
+                approximated_sm_efficiency = min(last_value, 1.0)
+                merged_ranges.append(((last_endpoint_time, ep.time), approximated_sm_efficiency))
+            last_endpoint_time = ep.time
+            if ep.pt_type == EndpointTypes.START:
+                last_value += ep.value
+            else:
+                last_value -= ep.value
+
+    return merged_ranges
+
+
+def merge_ranges(src_ranges, is_sorted=False):
+    merged_ranges = []
+    if len(src_ranges) > 0:
+        if not is_sorted:
+            src_ranges.sort(key=lambda x: x[0])
+        src_id = 0
+        merged_ranges.append(
+            (src_ranges[src_id][0], src_ranges[src_id][1]))
+        for src_id in range(1, len(src_ranges)):
+            dst_id = len(merged_ranges) - 1
+            if src_ranges[src_id][1] > merged_ranges[dst_id][1]:
+                if src_ranges[src_id][0] <= merged_ranges[dst_id][1]:
+                    merged_ranges[dst_id] = (merged_ranges[dst_id][0], src_ranges[src_id][1])
+                else:
+                    merged_ranges.append(
+                        (src_ranges[src_id][0], src_ranges[src_id][1]))
+    return merged_ranges
+
+
+def pop_list(range_list, index):
```

Comment:
wrong merge since:
1. the data.py is still use old code
2. most of the logic in this file is already in range_utils.py

Reply:
Thanks. Fixed.